CONTRIBUTING.md

```diff
@@ -800,45 +744,280 @@ Try restricting ifdefs in the header files, or in the beginning of
+Sometimes we make more commits than necessary, and we want to squash them
+(combine them). A good tool for this is _interactive rebase_.
+
+On your PR branch, run:
+
+```
+$ git rebase -i upstream/master
+```
+
+This will open an editor (usually `vi` or `emacs`), with a list of your
+commits:
+
+```
+pick ac07e6d27 Added new policy function isrealdir()
+pick 001bd420b Fixed mistake in isrealdir()
+
+# Commented out instructions ....
+```
+
+The commits / commands will be applied to `upstream/master` from top to bottom
+of this list. To combine the 2 commits, replace `pick` with `squash` on the
+second line:
+
+```
+pick ac07e6d27 Added new policy function isrealdir()
+squash 001bd420b Fixed mistake in isrealdir()
+
+# Commented out instructions ....
+```
+
+This will squash the second commit into the first one (the one above).
+Save and close. Another editor will open, allowing you to update the commit
+message(s). Update it to reflect the current state of this commit, then save
+and close.
+
+Once you get comfortable with _interactive rebase_ you can make more advanced
+changes, like squashing multiple commits, reordering commits, or dropping
+some commits. You may create conflicts, that you have to resolve, as
+described above.
+
+See also:
+[Squashing GitHub Pull Requests into a single commit](https://eli.thegreenplace.net/2014/02/19/squashing-github-pull-requests-into-a-single-commit)
+
+### Keeping master up to date
+
+Switch to local master branch:
+
+```
+$ git checkout master
+```
+
+Download changes from all remotes:
+
+```
+$ git fetch --all
+```
+
+Apply changes from the `upstream` `master` branch to our local `master` branch:
+
+```
+$ git rebase upstream/master
+```
+
+(You can also use `git pull`, but I find `rebase` a little easier when there
+are conflicts. `git pull` is a combination of `git fetch` and `git merge`.)
+
+Inspect the results:
+
+```
+$ git log
+```
+
+(Press `q` to quit `git log`).
+
+The latest (top) commit, should include `master` and `upstream/master` in
+parenthesis. This indicates that your local branch is up to date with
+`upstream`. If there are additional commits after the `upstream/master`
+commit, it means you committed some things on master, perhaps by accident.
+You can remove them, by doing:
+
+```
+$ git branch master-with-extra-commits
+$ git reset --hard upstream/master
+```
+
+You can, optionally, push your updated `master` branch to your GitHub fork:
+
+```
+$ git push
+```
+
+If you've already pushed some of the extra commits (on accident), you may
+have to add `--force`:
+
+```
+$ git push --force
+```
+
+However, normally this should not be necessary (!).
+
+
+### Updating git submodule(s)
+
+`git` has some commands for working with submodules, for example:
+
+```
+$ git submodule update
+```
+
+This will update the submodule, checking out the correct commit for your
+current `HEAD`.
+
+Please note that a submodule is just a normal git repo, with the `SHA`
```

Comment:
I wish this was true. Working in the submodule has proven to be impossible for me. I had to create a separate clone of the repo and work there. In particular, the submodule doesn't have its own `.git/config` and many things work in different ways.

Reply:
I have no problems working inside the submodule. I never interact with `.git/config` directly, but the config options I set using `git config` work just fine. Everything I've described in this guide, I've been able to do inside the submodule, in both this project and other projects which use submodules.

Reply:
Updated the wording here as well.

Reply:
Updated the wording here as well.